src/providers/pkcs11_provider/utils.rs

```diff
@@ -1,14 +1,274 @@
+            Hash::Sha384 => Ok(CkRsaPkcsMgfType::CkgMgf1Sha384),
+            Hash::Sha512 => Ok(CkRsaPkcsMgfType::CkgMgf1Sha512),
+            _ => Err(ResponseStatus::PsaErrorNotSupported),
+        }
+    }
+}
+
+impl TryFrom<Algorithm> for CkMechanism {
+    type Error = ResponseStatus;
+
+    fn try_from(alg: Algorithm) -> Result<Self> {
+        match alg {
+            Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPkcs1v15Sign { .. })
+            | Algorithm::AsymmetricEncryption(AsymmetricEncryption::RsaPkcs1v15Crypt { .. }) => {
+                Ok(CkMechanism::CkmRsaPkcs)
+            }
+            Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPss {
+                hash_alg: SignHash::Specific(hash_alg),
+            }) => Ok(CkMechanism::CkmRsaPkcsPss(CkRsaPkcsPssParams {
+                hash_alg: CkMechanism::try_from(hash_alg)?.mech_type(),
+                mgf: hash_alg.try_into()?,
+                s_len: hash_alg.hash_length(),
+            })),
+            Algorithm::AsymmetricEncryption(AsymmetricEncryption::RsaOaep { hash_alg }) => {
+                Ok(CkMechanism::CkmRsaPkcsOaep(CkRsaPkcsOaepParams {
+                    hash_alg: CkMechanism::try_from(hash_alg)?.mech_type(),
+                    mgf: hash_alg.try_into()?,
+                    source: CkRsaPkcsOaepSourceType::CkzDataSpecified,
+                    source_data: Vec::new(),
+                }))
+            }
+            _ => Err(ResponseStatus::PsaErrorNotSupported),
+        }
+    }
+}
+
+impl CkMechanism {
+    pub fn as_c_type(&mut self) -> (CK_MECHANISM, Option<CParams>) {
+        match self {
+            CkMechanism::CkmRsaPkcs
+            | CkMechanism::CkmSha1
+            | CkMechanism::CkmSha256
+            | CkMechanism::CkmSha384
+            | CkMechanism::CkmSha512 => (
+                CK_MECHANISM {
+                    mechanism: self.mech_type().into(),
+                    pParameter: std::ptr::null_mut(),
+                    ulParameterLen: 0,
+                },
+                None,
+            ),
+            CkMechanism::CkmRsaPkcsPss(ref params) => {
+                let mut params = params.as_c_type();
+                let len = ::std::mem::size_of_val(&params);
+                let p_params: CK_RSA_PKCS_PSS_PARAMS_PTR = &mut params;
+                (
+                    CK_MECHANISM {
+                        mechanism: self.mech_type().into(),
+                        pParameter: p_params as CK_VOID_PTR,
+                        ulParameterLen: len,
+                    },
+                    Some(CParams::CkmRsaPkcsPssParams(params)),
+                )
+            }
+            CkMechanism::CkmRsaPkcsOaep(ref mut params) => {
+                let mut params = params.as_c_type();
+                let len = ::std::mem::size_of_val(&params);
+                //ERROR: that does not work because the address of params might change when it is
```

Comment:
See here.

I thought that the address of Rust object would not change when they are moved which is why I return from this function the C `*_PARAMS` structure. But the address does change 😢

Reply:
ok, this was a fun one... 

I modified your code and now it works, will push in the morning. The way I did it was with `Pin`. Instead of holding the C `*_PARAMS` structure directly in `CParams`, I changed it to store `Pin<Box<*_PARAMS>>`. That way, the address is kept intact.

HOWEVER... my changes still didn't work initially and I spent a lot of time trying to figure out what was wrong... only to find that [SoftHSMv2 only supports SHA1 for OAEP](https://github.com/opendnssec/SoftHSMv2/issues/474) -_- Will have to decide if we should create some new tests, or if we want to implement support for the other algos in SoftHSM.

Reply:
(I have to say, though, this is rather convoluted... would be nice if we didn't have to do this in our code 😉)

Reply:
oh wow excellent idea! I finally understand what `Pin` is made for 👍

Reply:
I should remove that comment 😅